File: features/interpretation.ts
import type {Card, LanguagePreference, Reading, Spread, SpreadPosition} from '../entities';
import {SPREADS} from '../data';
import {I18n} from '../i18n';
import {findCardById, loadDeck} from '../utils/decks';

type Entry = {
    card: Card;
    position: SpreadPosition;
    isReversed: boolean;
};

type Translator = (key: string, vars?: Record<string, any>) => string;

export type Interpretation = {
    summary: string;
    positions: Array<{
        title: string;
        description: string;
        card: Card;
        isReversed: boolean;
        narrative: string;
    }>;
    keywords: string[];
};

const defaultTranslate: Translator = (key) => key;

const buildNarrative = (entry: Entry, t: Translator): string => {
    const cardSide = entry.isReversed ? entry.card.reversed : entry.card.upright;
    const positionLabel = t(entry.position.titleKey);
    const positionDescription = t(entry.position.descriptionKey);
    return `${positionLabel}: ${positionDescription} — ${cardSide.general}`;
};

const collectKeywords = (entries: Entry[]): string[] => {
    const keywords = entries.flatMap((entry) =>
        entry.isReversed ? entry.card.reversed.keywords : entry.card.upright.keywords,
    );
    return Array.from(new Set(keywords)).slice(0, 12);
};

export const hasBrokenSummary = (text: string | undefined): boolean =>
    Boolean(text && /\{(?:count|spread|price|model)\}/.test(text));

export const getReadingSummary = (reading: Reading, language: LanguagePreference): string => {
    if (reading.aiInsights?.summary) {
        return reading.aiInsights.summary;
    }

    const spread = SPREADS.find((item) => item.id === reading.spreadId);
    if (!spread) {
        return reading.summaryText ?? '';
    }

    const deck = loadDeck(language);
    const entries = reading.items
        .map((item) => {
            const position = spread.positions.find((pos) => pos.index === item.positionIndex);
            const card = findCardById(deck, item.cardId);
            if (!position || !card) return null;
            return {position, card, isReversed: item.isReversed};
        })
        .filter((entry): entry is Entry => entry !== null);

    if (!entries.length) {
        return reading.summaryText ?? '';
    }

    const translate: Translator = (key, vars) => I18n.t(key, {lng: language, ...(vars ?? {})});
    return generateInterpretation(spread, entries, translate).summary;
};

export const generateInterpretation = (
    spread: Spread,
    entries: Entry[],
    translate: Translator = defaultTranslate,
): Interpretation => {
    const t = translate;
    const positions = entries.map((entry) => ({
        title: t(entry.position.titleKey),
        description: t(entry.position.descriptionKey),
        card: entry.card,
        isReversed: entry.isReversed,
        narrative: buildNarrative(entry, t),
    }));

    const keywords = collectKeywords(entries);
    const spreadName = t(spread.nameKey);
    const intro = t('interpretation.summaryIntro', {
        spread: spreadName,
        count: entries.length,
    });
    const highlights = positions
        .slice(0, 3)
        .map((item) => item.narrative)
        .join(' ');
    const closing = t('interpretation.summaryClosing');

    return {
        summary: `${intro} ${highlights} ${closing}`.trim(),
        positions,
        keywords,
    };
};

File: ios-swift/MoonCompass/Core/Tarot/Interpretation.swift
import Foundation

/// A card placed in a spread position.
struct DrawnCard: Hashable, Sendable {
    let card: TarotCard
    let position: SpreadPosition
    let isReversed: Bool
}

/// The built-in, offline reading summary — the same text the React Native
/// client's `generateInterpretation` produces.
enum Interpretation {
    /// "Past: what shaped you — New beginnings, spontaneity…"
    ///
    /// Unlike the React Native text, each narrative ends with a full stop:
    /// card meanings have none, so consecutive positions used to run together
    /// ("…authority, truth Present: …").
    static func narrative(for entry: DrawnCard, localizer l: Localizer) -> String {
        var meaning = entry.card.meaning(reversed: entry.isReversed)
        if let last = meaning.last, !".!?…。".contains(last) {
            meaning += "."
        }
        return "\(l.t(entry.position.titleKey)): \(l.t(entry.position.descriptionKey)) — \(meaning)"
    }

    /// An intro, the first three positions' narratives and a closing line.
    static func summary(spread: Spread, entries: [DrawnCard], localizer l: Localizer) -> String {
        let intro = l.t("interpretation.summaryIntro", ["spread": l.t(spread.nameKey), "count": entries.count])
        let highlights = entries.prefix(3).map { narrative(for: $0, localizer: l) }.joined(separator: " ")
        let closing = l.t("interpretation.summaryClosing")
        return [intro, highlights, closing]
            .filter { !$0.isEmpty }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespaces)
    }

    /// Resolves a saved reading's cards against the spread and the catalog.
    /// Cards that can no longer be resolved are skipped rather than failing
    /// the whole reading.
    static func entries(for reading: Reading, spread: Spread, catalog: CardCatalog) -> [DrawnCard] {
        reading.items.compactMap { item in
            guard let position = spread.position(index: item.positionIndex),
                  let card = catalog.card(id: item.cardId) else { return nil }
            return DrawnCard(card: card, position: position, isReversed: item.isReversed)
        }
    }

    /// What to show for a reading in lists: the AI summary when there is one,
    /// otherwise the offline summary regenerated in the current language.
    static func displaySummary(for reading: Reading, catalog: CardCatalog, localizer l: Localizer) -> String {
        if let ai = reading.aiInsights?.summary, !ai.isEmpty {
            return ai
        }
        guard let spread = Spread.find(reading.spreadId) else { return reading.summaryText }
        let entries = entries(for: reading, spread: spread, catalog: catalog)
        guard !entries.isEmpty else { return reading.summaryText }
        return summary(spread: spread, entries: entries, localizer: l)
    }
}
